refactor: replace `simplify_optiontype` and `simplify_uniontype` with `simplified` classmethod

jpivarski opened the pull request:
Instead of constructing non-canonical arrays and then simplifying them into something canonical, these `@classmethods` only construct canonical arrays from the start.

The next PR after this will enforce canonicity (level 1) in the Content constructors. This is the first half of #1910.

<!-- docs-preview-start -->
----
:books: The documentation for this PR will be available at <https://awkward-array.readthedocs.io/en/jpivarski-replace-simplify-method-with-classmethod-constructor/> once Read the Docs has finished building :hammer:
<!-- docs-preview-end -->

jpivarski committed:
Added @classmethod simplified (and copy) to all Forms.

jpivarski committed:
Is adding a 'simplified' option to to_IndexedArray64 right?

jpivarski committed:
No, I don't think so.

jpivarski committed:
Merge branch 'main' into jpivarski/replace-simplify-method-with-classmethod-constructor

jpivarski committed:
Implemented, but have not tested, 'simplified' classmethods.

agoose77 commented on src/awkward/forms/bitmaskedform.py:
```diff
@@ -1,12 +1,44 @@
+            self._valid_when if valid_when is unset else valid_when,
+            self._lsb_order if lsb_order is unset else lsb_order,
+            parameters=self._parameters if parameters is unset else parameters,
+            form_key=self._form_key if form_key is form_key else form_key,
```
```suggestion
            form_key=self._form_key if form_key is unset else form_key,
```

Not sure if you're taking review at draft stage, but I noticed this and flagged it here :)

jpivarski replied:
Got it, thanks!

agoose77 commented on src/awkward/forms/form.py:
```diff
@@ -412,3 +413,14 @@ def _to_dict_part(self, verbose, toplevel):
     def _type(self, typestrs):
         raise _errors.wrap_error(NotImplementedError)
+
+    def length_zero_array(self, *, nplike=numpy, highlevel=True, behavior=None):
```
Yes! It now has a name!

jpivarski committed:
Every Content and Form should have a 'simplified' constructor, and they're right after the normal constructor, now.

jpivarski committed:
Removed all calls of 'simplify_optiontype' (except within 'simplify_optiontype').

jpivarski commented on src/awkward/_slicing.py:
```diff
@@ -266,8 +266,7 @@ def normalise_item_nested(item):
     ):
-        next = item.simplify_optiontype()
-        return normalise_item_nested(next)
+        return normalise_item_nested(item)
```
This is one of the few that doesn't replace a `simplify_optiontype` with a `simplified` classmethod—not a pure refactor—so it's possible that non-canonical layouts pass through here.

However, I think this was sanitizing user input, and after constructors forbid non-canonical structures, user input can't be insane. So while this opens the door to error, the next PR will shut it.

jpivarski commented on src/awkward/contents/bitmaskedarray.py:
```diff
@@ -632,7 +632,7 @@ def _completely_flatten(self, backend, options):
         if branch or options["drop_nones"] or depth > 1:
             return self.project()._completely_flatten(backend, options)
         else:
-            return [self.simplify_optiontype()]
+            return [self]
```
Here's another unbalanced `simplify_optiontype` → `simplified` pair, but in the future, `self` can't be non-canonical.

jpivarski commented on src/awkward/contents/bytemaskedarray.py:
```diff
@@ -1019,7 +1015,7 @@ def _completely_flatten(self, backend, options):
         if branch or options["drop_nones"] or depth > 1:
             return self.project()._completely_flatten(backend, options)
         else:
-            return [self.simplify_optiontype()]
+            return [self]
```
Another unbalanced `simplify_optiontype` → `simplified` pair.

jpivarski commented on src/awkward/contents/indexedoptionarray.py:
```diff
@@ -1589,7 +1588,7 @@ def _completely_flatten(self, backend, options):
             return self.project()._completely_flatten(backend, options)
         else:
-            return [self.simplify_optiontype()]
+            return [self]
```
Another unbalanced `simplify_optiontype` → `simplified` pair.

jpivarski commented on src/awkward/contents/unmaskedarray.py:
```diff
@@ -479,7 +476,7 @@ def _completely_flatten(self, backend, options):
             return self.project()._completely_flatten(backend, options)
         else:
-            return [self.simplify_optiontype()]
+            return [self]
```
Another unbalanced `simplify_optiontype` → `simplified` pair.

jpivarski commented on src/awkward/operations/ak_nan_to_none.py:
```diff
@@ -46,7 +46,7 @@ def action(layout, continuation, **kwargs):
             and issubclass(layout.content.dtype.type, np.floating)
         ):
-            return continuation().simplify_optiontype()
+            return continuation()
```
Another unbalanced `simplify_optiontype` → `simplified` pair.

jpivarski commented on src/awkward/operations/ak_to_categorical.py:
```diff
@@ -83,8 +83,6 @@ def to_categorical(array, *, highlevel=True, behavior=None):
 def _impl(array, highlevel, behavior):
     def action(layout, **kwargs):
         if layout.purelist_depth == 1:
-            if layout.is_option:
-                layout = layout.simplify_optiontype()
```
The last unbalanced `simplify_optiontype` → `simplified` pair.

agoose77 committed:
fix: use `backend` instead of `nplike`

This commit enables backends to pass through `regularize_backend`, and
ensures that we pass a backend (not nplike) to `from_buffers`

agoose77 commented on src/awkward/_backends.py:
```diff
@@ -189,8 +189,10 @@ def backend_of(*objects, default: D = _UNSET) -> Backend | D:
 
-def regularize_backend(backend: str) -> Backend:
-    if backend in _backends:
+def regularize_backend(backend: str | Backend) -> Backend:
```
I haven't changed the type signature for `ak.from_buffers` to reflect the ability to pass an existing `Backend`. We could choose to do that later, but for now I think the string name should be the user-public API.

This change just ensures that internally we can call `ak_from_buffers._impl` with an existing `Backend` object.

jpivarski committed:
Removed all calls of 'simplify_uniontype' (except within 'simplify_uniontype').

jpivarski commented on src/awkward/_slicing.py:
```diff
@@ -322,15 +322,11 @@ def normalise_item_nested(item):
+        raise ak._errors.wrap_error(
+            TypeError(
+                "irreducible unions (different types at the same level in an array) can't be used as slices"
             )
-
-        return normalise_item_nested(attempt)
+        )
```
This one is not a `simplify_uniontype` → `simplified` pair because it's sanitizing input, just in case (like the option-type exceptions noted above).

jpivarski commented on src/awkward/contents/unionarray.py:
```diff
@@ -189,7 +190,7 @@ def simplified(
         contents = []
 
-        for i, self_cont in enumerate(contents):
+        for i, self_cont in enumerate(self_contents):
```
This had been an error in the translation of `simplify_uniontype` to `simplified`. I don't see any similar errors.

jpivarski commented on src/awkward/contents/unionarray.py:
```diff
@@ -614,18 +615,15 @@ def _nested_tags_index(self, offsets: Index, counts: Sequence[Index]):
-        if isinstance(simplified, ak.contents.UnionArray):
+        if isinstance(self, ak.contents.UnionArray):
             raise ak._errors.index_error(
                 self,
                 ak.contents.ListArray(
                     slicestarts, slicestops, slicecontent, parameters=None
                 ),
                 "cannot apply jagged slices to irreducible union arrays",
             )
-        return simplified._getitem_next_jagged(
-            slicestarts, slicestops, slicecontent, tail
-        )
+        return self._getitem_next_jagged(slicestarts, slicestops, slicecontent, tail)
```
This one is not a `simplify_uniontype` → `simplified` pair because it's sanitizing input. There was a test providing insane inputs, so the `simplified` part of the pair went into the testing code.

jpivarski commented on src/awkward/contents/unionarray.py:
```diff
@@ -1293,7 +1292,15 @@ def numbers_to_type(self, name):
         )
 
     def _is_unique(self, negaxis, starts, parents, outlength):
-        simplified = self.simplify_uniontype(True, True)
+        simplified = type(self).simplified(
+            self._tags,
+            self._index,
+            self._contents,
+            parameters=self._parameters,
+            backend=self._backend,
+            merge=True,
+            mergebool=True,
+        )
```
This one re-simplifies the inputs because the merging strategy is more aggressive than normal (`mergebool=True`), so this is to prevent any changes in behavior.

jpivarski commented on src/awkward/contents/unionarray.py:
```diff
@@ -1302,7 +1309,15 @@ def _is_unique(self, negaxis, starts, parents, outlength):
         return simplified._is_unique(negaxis, starts, parents, outlength)
 
     def _unique(self, negaxis, starts, parents, outlength):
-        simplified = self.simplify_uniontype(True, True)
+        simplified = type(self).simplified(
+            self._tags,
+            self._index,
+            self._contents,
+            parameters=self._parameters,
+            backend=self._backend,
+            merge=True,
+            mergebool=True,
+        )
```
See above.

jpivarski commented on src/awkward/contents/unionarray.py:
```diff
@@ -1322,7 +1337,14 @@ def _argsort_next(
         kind,
         order,
     ):
-        simplified = self.simplify_uniontype(mergebool=True)
+        simplified = type(self).simplified(
+            self._tags,
+            self._index,
+            self._contents,
+            parameters=self._parameters,
+            backend=self._backend,
+            mergebool=True,
+        )
```
See above.

jpivarski commented on src/awkward/contents/unionarray.py:
```diff
@@ -1345,7 +1367,14 @@ def _sort_next(
         if self.length == 0:
             return self
 
-        simplified = self.simplify_uniontype(mergebool=True)
+        simplified = type(self).simplified(
+            self._tags,
+            self._index,
+            self._contents,
+            parameters=self._parameters,
+            backend=self._backend,
+            mergebool=True,
+        )
```
See above.

jpivarski commented on src/awkward/contents/unionarray.py:
```diff
@@ -1370,7 +1399,14 @@ def _reduce_next(
         keepdims,
         behavior,
     ):
-        simplified = self.simplify_uniontype(mergebool=True)
+        simplified = type(self).simplified(
+            self._tags,
+            self._index,
+            self._contents,
+            parameters=self._parameters,
+            backend=self._backend,
+            mergebool=True,
+        )
```
See above.

jpivarski commented on src/awkward/operations/ak_concatenate.py:
```diff
@@ -115,7 +115,15 @@ def _impl(arrays, axis, merge, mergebool, highlevel, behavior):
         out = batch[0].mergemany(batch[1:])
         if isinstance(out, ak.contents.UnionArray):
-            out = out.simplify_uniontype(merge=merge, mergebool=mergebool)
+            out = type(out).simplified(
+                out._tags,
+                out._index,
+                out._contents,
+                parameters=out._parameters,
+                backend=out._backend,
+                merge=merge,
+                mergebool=mergebool,
+            )
```
See above (`mergebool` _might_ be `True`).

jpivarski committed:
Merge branch 'main' into jpivarski/replace-simplify-method-with-classmethod-constructor

jpivarski committed:
Redundant statements of the defaults.

jpivarski committed:
Removed 'simplify_optiontype' and 'simplify_uniontype'.

jpivarski committed:
Use 'cls' where possible (one of which fixed an untested bug).

jpivarski committed:
The 'simplified' classmethod also prevents option-of-union now.

agoose77 commented on src/awkward/contents/unionarray.py:
```diff
@@ -282,6 +477,82 @@ def _carry(self, carry, allow_lazy):
+    def _union_of_optionarrays(self, index, parameters):
+        tag_for_missing = 0
+        for i, content in enumerate(self._contents):
+            if content.is_record:
```
```suggestion
            if content.is_option:
```

agoose77 commented on src/awkward/contents/unionarray.py:
```diff
@@ -282,6 +477,82 @@ def _carry(self, carry, allow_lazy):
+            contents = []
+            for tag, content in enumerate(self._contents):
+                if tag == tag_for_missing:
+                    indexedoption_index = self._backend.nplike.arange(
```
```suggestion
                    indexedoption_index = self._backend.index_nplike.arange(
```

agoose77 commented on src/awkward/contents/unionarray.py:
```diff
@@ -282,6 +477,82 @@ def _carry(self, carry, allow_lazy):
+                nexttags = self._backend.nplike.full(
+                    len(carry_data), tag_for_missing, dtype=self._tags.dtype
+                )
+                nextindex = self._backend.nplike.full(
```
```suggestion
                nextindex = self._backend.index_nplike.full(
```

agoose77 commented on src/awkward/contents/unionarray.py:
```diff
@@ -282,6 +477,82 @@ def _carry(self, carry, allow_lazy):
+
+            else:
+                # UnionArray is empty, so
+                nexttags = self._backend.nplike.full(
```
```suggestion
                nexttags = self._backend.index_nplike.full(
```

agoose77 commented on src/awkward/contents/unionarray.py:
```diff
@@ -282,6 +477,82 @@ def _carry(self, carry, allow_lazy):
+            contents = []
+            for tag, content in enumerate(self._contents):
+                if tag == tag_for_missing:
+                    indexedoption_index = self._backend.nplike.arange(
```
```suggestion
                    indexedoption_index = self._backend.index_nplike.arange(
```

agoose77 commented on src/awkward/operations/ak_where.py:
```diff
@@ -122,7 +122,16 @@ def action(inputs, **kwargs):
             if not isinstance(right, ak.contents.Content):
                 right = ak.contents.NumpyArray(backend.nplike.repeat(right, len(tags)))
             tmp = ak.contents.UnionArray(tags, index, [left, right])
```
Let's replace this with a single `simplified` call?

jpivarski replied:
Fixed in fb68bfd308467e43af0dab240fde36126cc018af.

jpivarski committed:
Fixed unnecessary double-construction.

jpivarski committed:
nplike → index_nplike

Co-authored-by: Angus Hollands <goosey15@gmail.com>

jpivarski committed:
nplike → index_nplike

Co-authored-by: Angus Hollands <goosey15@gmail.com>

jpivarski committed:
nplike → index_nplike

Co-authored-by: Angus Hollands <goosey15@gmail.com>

jpivarski committed:
nplike → index_nplike

Co-authored-by: Angus Hollands <goosey15@gmail.com>

jpivarski committed:
Choose `tag_for_missing` based on `is_option`, rather than `is_record`.

Co-authored-by: Angus Hollands <goosey15@gmail.com>
